### PR TITLE
Finish off IPAM TODOs

### PR DESCRIPTION
**Pull request by caseydavenport**
Includes a few fixes each in its own commit:
- [x] RemoveIPAMHost implementation
- [x] Minimize datastore writes when releasing IP address
- [x] Implement random block hunt when out of free blocks

### CODE REVIEW

File: lib/client/ipam.go [hunk 1]
```diff
@@ -589,22 +655,18 @@ func (c ipams) RemoveIPAMHost(host *string) error {
...
-	// 		return err
-	// 	}
-	// }
+	// Remove the host tree from the datastore.
+	err := c.client.backend.Delete(&model.KVPair{
+		Key: model.IPAMHostKey{Host: hostname},
+	})
+	if err != nil {
+		if _, ok := err.(common.ErrorResourceDoesNotExist); ok {
+			// Already deleted - carry on.
```
Comment (robbrockbank) [inline lib/client/ipam.go]: contract to a single branch?   !ok

Comment (caseydavenport) [inline lib/client/ipam.go]: d'oh, yeah...

File: lib/client/ipam.go [hunk 2]
```diff
@@ -589,22 +655,17 @@ func (c ipams) RemoveIPAMHost(host *string) error {
 	// Determine the hostname to use.
 	hostname := decideHostname(host)
 
-	// Release host affinities.
```
Comment (robbrockbank) [inline lib/client/ipam.go]: Did you mean to remove this as well?

Comment (caseydavenport) [inline lib/client/ipam.go]: Well, I could leave it.  However, removing the IPAM host actually removes all the affinities anyway so I took it out.

I'm not a super big fan of it, but I suppose it's sort of like removing a tier and thus it removes all the policies within the tier with it.

**Comment by robbrockbank** [comment]
One comment - looks like you removed some code that I don't think should have been removed.

File: lib/client/ipam.go (hunk 2 above)
Comment (robbrockbank) [inline lib/client/ipam.go]: I thought it did more than just that - I thought it reconfigured the allocation block to remove the host affinity.

Comment (caseydavenport) [inline lib/client/ipam.go]: Right you are.

**Comment by caseydavenport** [comment]
OK - comment addressed!

Good to go?

**Comment by robbrockbank** [comment]
LGTM

**Pull request merged**